File: 백준/Gold/1005. ACM Craft/ACM Craft.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int t;

void topologySort(int n, int k, vector<int>& inDegree, vector<int>& res, vector<vector<int>>& graph, int w) {
	queue<int> q;
	vector<int> dp(n+1);
	for (int i = 1; i < n+1; i++) {
		if (inDegree[i] == 0) {
			q.push(i);
			dp[i] = res[i];
		}
	}

	while (!q.empty()) {
		int node = q.front();
		q.pop();

		for (int i = 0; i < graph[node].size(); i++) {
			int nNode = graph[node][i];
			
			inDegree[nNode]--;
			dp[nNode] = max(dp[nNode], dp[node]);
			if (inDegree[nNode] == 0) {
				q.push(nNode);
				dp[nNode] += res[nNode];
			}
		}

		if (inDegree[w] == 0) {
			cout << dp[w] << "\n";

			return;
		}

	}

}

// 사이클 없음
int main() {
	ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

	cin >> t;
	while(t--) {
		int n, k, w;
		cin >> n >> k;
		vector<int> inDegree(n+1);
		vector<int> res(n+1);
		vector<vector<int>> graph(n+1);
		for (int i = 1; i < n+1; i++) cin >> res[i];
		for (int i = 0; i < k; i++) {
			int x, y;
			cin >> x >> y;

			graph[x].push_back(y);
			inDegree[y]++;
		}

		cin >> w;

		topologySort(n, k, inDegree, res, graph, w);
	}
}

File: 백준/Gold/1005. ACM Craft/ACM Craft.java
import java.io.*;
import java.util.*;

public class Main {
    static int t, n, k, w;
    static int[] res, inDegree;
    static ArrayList<ArrayList<Integer>> graph;

    private static void topologySort(StringBuilder sb) {
        int[] answer = new int[n+1];

        Deque<Integer> q = new ArrayDeque<>();
        for (int i = 1; i <= n; i++) {
            if (inDegree[i] == 0) {
                q.offer(i);
                answer[i] = res[i];
            }
        }

        while (!q.isEmpty()) {
            int node = q.poll();

            for (int next : graph.get(node)) {
                if (--inDegree[next] == 0) {
                    q.offer(next);
                }
                answer[next] = Math.max(answer[next], (answer[node] + res[next]));
            }
        }

        sb.append(answer[w]).append("\n");
    }

    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            n = Integer.parseInt(st.nextToken());
            k = Integer.parseInt(st.nextToken());

            graph = new ArrayList<>();
            for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());

            inDegree = new int[n+1];
            res = new int[n+1];
            st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= n; i++) res[i] = Integer.parseInt(st.nextToken());

            for (int i = 0; i < k; i++) {
                int a, b;
                st = new StringTokenizer(br.readLine());

                a = Integer.parseInt(st.nextToken());
                b = Integer.parseInt(st.nextToken());

                graph.get(a).add(b);
                inDegree[b]++;
            }

            w = Integer.parseInt(br.readLine());
            topologySort(sb);
        }
        System.out.println(sb);
    }
}
